flink-runtime-web/src/main/java/org/apache/flink/runtime/webmonitor/metrics/MetricStore.java

```diff
@@ -0,0 +1,172 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.flink.runtime.webmonitor.metrics;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.HashMap;
+import java.util.Map;
+
+/**
+ * Nested data-structure to store metrics.
+ *
+ * This structure is not thread-safe.
+ */
+public class MetricStore {
+	private static final Logger LOG = LoggerFactory.getLogger(MetricStore.class);
+
+	JobManagerMetricStore jobManager = new JobManagerMetricStore();
+	Map<String, TaskManagerMetricStore> taskManagers = new HashMap<>();
+	Map<String, JobMetricStore> jobs = new HashMap<>();
+
+	/**
+	 * Adds a metric to this MetricStore.
+	 *
+	 * @param name  the metric identifier
+	 * @param value the metric value
+	 */
+	public void add(String name, Object value) {
```

Comment:
I think we are mixing 2 separate issues here:
A) the format of the transfer
B) how the retrieved metrics are added to the store

Your comment regarding reparsing the string for Histograms targets B. This can be fixed easily by providing specific `addHistogram/...` methods to the store.

Now, let's talk about A. Essentially, the goal was to reduce the overhead for the sender. If you look at the dumping procedure in the `MetricQueryService` you will find that for the sender the current requires the least amount of work, as far as i can tell. We don't create more elaborate containers (which btw. i would personally prefer too!) as this is another short-lived object created for every metric.

Also, in regards to sender/receiver going out of sync: we can add a test to guarantee that sender&receiver do not go out of sync.